stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -2,29 +2,108 @@
 
 ## linalg.jl: Some generic Linear Algebra definitions
 
-function generic_mul!(C::AbstractArray, X::AbstractArray, s::Number)
+"""
+    MulAddMul(alpha, beta)
+
+A callable for operating short-circuiting version of `x * alpha + y * beta`.
+
+# Examples
+```jldoctest
+julia> using LinearAlgebra: MulAddMul
+
+julia> _add = MulAddMul(1, 0);
+
+julia> _add(123, nothing)
+123
+
+julia> MulAddMul(12, 34)(56, 78) == 56 * 12 + 78 * 34
+true
+```
+"""
+struct MulAddMul{ais1, bis0, TA, TB}
+    alpha::TA
+    beta::TB
+end
+
+MulAddMul(alpha::TA, beta::TB) where {TA, TB} =
+    MulAddMul{isone(alpha), iszero(beta), TA, TB}(alpha, beta)
```

Comment:
It might be preferable for dispatch/inference to write this as
```
@inline function MulAddMul(alpha::TA, beta::TB) where {TA, TB}
    if isone(alpha)
        if iszero(beta)
           return MulAddMul{true, true, TA,TB}(alpha, beta)
       else
           return MulAddMul{true, false, TA, TB}(alpha, beta)
       end
       else
            if iszero(beta)
               return MulAddMul{false, true, TA,TB}(alpha, beta)
           else
                return MulAddMul{false, false, TA, TB}(alpha, beta)
           end
        end
       nothing
end
```
This would help the compiler/inference make an exhaustive list of cases instead of potentially open-ended dynamic dispatch.

Reply:
This is a very clever idea.  But, FYI, I think the benchmark I ran while ago https://github.com/JuliaLang/julia/pull/29634#issuecomment-440512432 suggests that constant propagation took care of it (at the cost of `@inline`ing functions).